migration/src/main/scala/akka/persistence/r2dbc/migration/DurableStateMigrationToolDao.scala

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (C) 2022 - 2023 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.persistence.r2dbc.migration
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+
+import io.r2dbc.spi.Connection
+
+import akka.Done
+import akka.annotation.InternalApi
+import akka.dispatch.ExecutionContexts
+import akka.persistence.Persistence
+import akka.persistence.query.NoOffset
+import akka.persistence.query.UpdatedDurableState
+import akka.persistence.r2dbc.internal.Dialect
+import akka.persistence.r2dbc.internal.DurableStateDao.EmptyDbTimestamp
+import akka.persistence.r2dbc.internal.DurableStateDao.SerializedStateRow
+import akka.persistence.r2dbc.internal.R2dbcExecutor
+import akka.persistence.r2dbc.internal.R2dbcExecutorProvider
+import akka.persistence.r2dbc.internal.Sql.InterpolationWithAdapter
+import akka.persistence.r2dbc.internal.codec.PayloadCodec.RichStatement
+import akka.persistence.r2dbc.internal.postgres.PostgresDurableStateDao
+import akka.persistence.r2dbc.internal.postgres.PostgresDurableStateDao.FutureDone
+import akka.persistence.typed.PersistenceId
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[r2dbc] class DurableStateMigrationToolDao(
+    persistenceExt: Persistence,
+    executorProvider: R2dbcExecutorProvider,
+    dialect: Dialect)(implicit ec: ExecutionContext)
+    extends PostgresDurableStateDao(executorProvider, dialect) {
+
+  import settings.codecSettings.DurableStateImplicits._
+
+  def upsertDurableState(state: SerializedStateRow, value: Any): Future[Done] = {
+    val slice = persistenceExt.sliceForPersistenceId(state.persistenceId)
+
+    val r2dbcExecutor = executorProvider.executorFor(slice)
+    val entityType = PersistenceId.extractEntityType(state.persistenceId)
+    val bindings = additionalBindings(entityType, state, value)
+
+    val upsertSql =
+      sql"""
+         ${insertStateSql(slice, entityType, bindings)}
+         ON CONFLICT(persistence_id)
+         DO UPDATE SET
+           slice = EXCLUDED.slice,
+           entity_type = EXCLUDED.entity_type,
+           revision = EXCLUDED.revision,
+           state_ser_id = EXCLUDED.state_ser_id,
+           state_ser_manifest = EXCLUDED.state_ser_manifest,
+           state_payload = EXCLUDED.state_payload,
+           tags = EXCLUDED.tags,
+           test_column = EXCLUDED.test_column,
+           db_timestamp = EXCLUDED.db_timestamp
```

Comment:
How does this work with additionalColumns? they are bound to the insertStateSql, but not part of those EXCLUDED updates.

If that is true I think it's a good indication that this is becoming too complex with risk of bugs and more difficult to maintain, and we should instead go back to #554

Reply:
Right, additional Columns (as well as change handler?) are missing here.